sdk/storage/azure-storage-blob/azure/storage/blob/_download.py

```diff
@@ -774,8 +816,65 @@ def readinto(self, stream: IO[bytes]) -> int:
             for chunk in downloader.get_chunk_offsets():
                 downloader.process_chunk(chunk)
 
+        self._complete_read()
         return remaining_size
 
+    def _complete_read(self):
+        """Adjusts all offsets to the end of the download."""
```

Comment:
I'm curious why this was needed? Is it just a side-effect of where the loops exit that mean these values don't necessarily reach "completion" once all the data is processed?

Reply:
This is used in the case where someone calls `readall` or `read` with no size specified (which can happen after partial stream read). In those cases, we use the chunk downloader directly which does not update any offsets in this class. This just sets everything to "done" to make sure subsequent calls short circuit and return empty data.